br/pkg/metautil/statsfile.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/pingcap/tidb/pkg/statistics/handle"
+	statstypes "github.com/pingcap/tidb/pkg/statistics/handle/types"
+	statsutil "github.com/pingcap/tidb/pkg/statistics/handle/util"
+	"golang.org/x/sync/errgroup"
+)
+
+var maxStatsJsonTableSize = 32 * 1024 * 1024 // 32 MiB
+var inlineSize = 8 * 1024                    // 8 KiB
+
+func getStatsFileName(physicalID int64) string {
+	return fmt.Sprintf("backupmeta.schema.stats.%09d", physicalID)
+}
+
+// A lightweight function wrapper to dump the statistic
+type StatsWriter struct {
+	storage storage.ExternalStorage
+	cipher  *backuppb.CipherInfo
+
+	// final stats file indexes
+	statsFileIndexes []*backuppb.StatsFileIndex
+
+	// temporary variables, clear after each flush
+	totalSize int
+	statsFile *backuppb.StatsFile
+}
+
+func newStatsWriter(
+	storage storage.ExternalStorage,
+	cipher *backuppb.CipherInfo,
+) *StatsWriter {
+	return &StatsWriter{
+		storage: storage,
+		cipher:  cipher,
+
+		statsFileIndexes: make([]*backuppb.StatsFileIndex, 0),
+
+		totalSize: 0,
+		statsFile: &backuppb.StatsFile{},
+	}
+}
+
+func (s *StatsWriter) writeStatsFileAndClear(ctx context.Context, physicalID int64) error {
+	fileName := getStatsFileName(physicalID)
+	content, err := proto.Marshal(s.statsFile)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if len(s.statsFileIndexes) == 0 && len(content) < inlineSize {
+		s.statsFileIndexes = append(s.statsFileIndexes, &backuppb.StatsFileIndex{InlineData: content})
+		return nil
+	}
+
+	checksum := sha256.Sum256(content)
+
+	encryptedContent, iv, err := Encrypt(content, s.cipher)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	if err := s.storage.WriteFile(ctx, fileName, encryptedContent); err != nil {
+		return errors.Trace(err)
+	}
+
+	s.statsFileIndexes = append(s.statsFileIndexes, &backuppb.StatsFileIndex{
+		Name:     fileName,
+		Sha256:   checksum[:],
+		SizeEnc:  uint64(len(encryptedContent)),
+		SizeOri:  uint64(len(content)),
+		CipherIv: iv,
+	})
+
+	// clear the temporary variables
+	s.totalSize = 0
+	s.statsFile = &backuppb.StatsFile{}
+	return nil
+}
+
+func (s *StatsWriter) BackupStats(ctx context.Context, jsonTable *statsutil.JSONTable, physicalID int64) error {
+	if jsonTable == nil {
+		return nil
+	}
+
+	statsBytes, err := json.Marshal(jsonTable)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	s.totalSize += len(statsBytes)
+	s.statsFile.Blocks = append(s.statsFile.Blocks, &backuppb.StatsBlock{
+		PhysicalId: physicalID,
+		JsonTable:  statsBytes,
+	})
+
+	// check whether need to flush
+	if s.totalSize > maxStatsJsonTableSize {
+		if err := s.writeStatsFileAndClear(ctx, physicalID); err != nil {
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (s *StatsWriter) BackupStatsDone(ctx context.Context) ([]*backuppb.StatsFileIndex, error) {
+	if s.totalSize == 0 || len(s.statsFile.Blocks) == 0 {
+		return s.statsFileIndexes, nil
+	}
+
+	if err := s.writeStatsFileAndClear(ctx, s.statsFile.Blocks[0].PhysicalId); err != nil {
```

Comment:
may be it's ok for now this file name does not duplicated.
what if file name duplicated? is it could happen during backup checkpoint?

Reply:
use physical id to distinguish files.